### PR TITLE
CORE-461 (This could work)

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: src/passphrase_cache.cc [hunk 1]
```diff
@@ -32,26 +37,47 @@ namespace pEp {
...
     }
 
+    const char *PassphraseCache::add(const cache_entry& entry)
+    {
+        const char* result = nullptr;
+        std::lock_guard<std::mutex> lock(_mtx);
+
+        auto found = find(_cache.begin(), _cache.end(), entry);
+
+        if (found != _cache.end()) {
```
Comment (ig-garcia) [inline src/passphrase_cache.cc:62]: I was wondering if we would replace the cache entry if there is one with same email address but different passphrase? or this is a case that cannot happen?

File: src/passphrase_cache.cc [hunk 2]
```diff
@@ -60,6 +86,12 @@ namespace pEp {
         return empty;
     }
 
+    void PassphraseCache::remove(const std::string& account_email, const std::string& passphrase)
```
Comment (ig-garcia) [inline src/passphrase_cache.cc]: Should we remove entries only by email? or we expect having several entries for same email address?

File: src/passphrase_cache.cc (hunk 1 above)
Comment (ghost) [reply src/passphrase_cache.cc:62]: Will insert checks so we uphold valid state.

File: src/passphrase_cache.cc (hunk 2 above)
Comment (ghost) [reply src/passphrase_cache.cc]: Will do.

**Comment by ghost** [comment]
I have made the behavior explicit. Please also see https://plancksecurity.atlassian.net/browse/CORE-461?focusedCommentId=44645.

File: src/passphrase_cache.cc [hunk 3]
```diff
@@ -45,10 +45,22 @@ namespace pEp {
         const char* result = nullptr;
         std::lock_guard<std::mutex> lock(_mtx);
 
-        auto found = find(_cache.begin(), _cache.end(), entry);
+        cache::iterator found = _cache.end();
+
+        if (entry.account_email.empty()) {
+            // search by passphrase
+            found = find_if(_cache.begin(), _cache.end(), [&entry](const cache_entry& e) {
```
Comment (ig-garcia) [inline src/passphrase_cache.cc:52]: Could this risk that we remove the email from an entry with email and convert it into an entry without email? should we maybe check only entries with empty email?

**Review by saschabacardit — APPROVED** [review]

File: src/passphrase_cache.cc (hunk 3 above)
Comment (ghost) [reply src/passphrase_cache.cc:52]: Updated.

**Review by ig-garcia — APPROVED** [review]

**Pull request merged**